design/extension.go

```diff
@@ -0,0 +1,963 @@
+
+	Method("listExtensions", func() {
+		Description("List extensions registered in FuseML")
+
+		Result(ArrayOf(Extension), "Return all registered extensions.")
+
+		HTTP(func() {
+			GET("/extensions")
+			Response(StatusOK)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+		})
+	})
+
+	Method("updateExtension", func() {
+		Description("Update an extension registered in FuseML")
+
+		Payload(Extension, "Extension update request")
+
+		Result(Extension, "Return the updated extension.")
+
+		Error("BadRequest", func() {
+			Description("If the extension does not have the required fields, should return 400 Bad Request.")
+		})
+
+		Error("NotFound", func() {
+			Description("If the extension is not found, should return 404 Not Found.")
+		})
+
+		HTTP(func() {
+			PUT("/extensions/{id}")
+			Response(StatusOK)
+			Response("NotFound", StatusNotFound)
+			Response("BadRequest", StatusBadRequest)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("deleteExtension", func() {
+		Description("Delete an extension and its subtree of services, endpoints and credentials")
+		Payload(func() {
+			Field(1, "id", String, "Extension identifier", func() {
+				Pattern(identifierPattern)
+				MaxLength(100)
+				Example("kfserving-001")
+			})
+			Required("id")
+		})
+
+		Error("BadRequest", func() {
+			Description("If the extension cannot be deleted, should return 400 Bad Request.")
+		})
+
+		Error("NotFound", func() {
+			Description("If the extension is not found, should return 404 Not Found.")
+		})
+
+		HTTP(func() {
+			DELETE("/extensions/{id}")
+			Response(StatusNoContent)
+			Response("NotFound", StatusNotFound)
+			Response("BadRequest", StatusBadRequest)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+		})
+	})
+
+	Method("addService", func() {
+		Description("Add a service to an existing extension registered with the FuseML extension registry.")
+
+		Payload(ExtensionService, "Extension service add request")
+
+		Error("NotFound", func() {
+			Description("If the extension is not found, should return 404 Not Found.")
+		})
+		Error("BadRequest", func() {
+			Description("If the service does not have the required fields, should return 400 Bad Request.")
+		})
+		Error("Conflict", func() {
+			Description("If an extension service with the same name already exists, should return 409 Conflict.")
+		})
+
+		Result(ExtensionService)
+
+		HTTP(func() {
+			POST("/extensions/{extension_id}/services")
+			Response(StatusCreated)
+			Response("NotFound", StatusNotFound)
+			Response("BadRequest", StatusBadRequest)
+			Response("Conflict", StatusConflict)
+		})
+
+		GRPC(func() {
+			Response(CodeOK)
+			Response("NotFound", CodeNotFound)
+			Response("BadRequest", CodeInvalidArgument)
+			Response("Conflict", CodeAlreadyExists)
+		})
+	})
+
+	Method("getService", func() {
```

Comment:
Shouldn't `getService` be just a special case of `listServices`, with `id` argument being optional? I think we have it similar for other resources

Reply:
No, they are two different things. getService targets a specific service and returns an error if the service isn't found. listServices would only return an empty list if the criteria isn't matching any of the services.

These REST operations also target different REST resources (look at the URL route), which is one more reason why they should be different commands.

Reply:
Ok